whisper.cpp

```diff
@@ -1636,24 +1630,25 @@ static struct ggml_cgraph * whisper_build_graph_encoder(
     const int n_head  = hparams.n_audio_head;
     const int n_layer = hparams.n_audio_layer;
 
+    static size_t buf_size = ggml_tensor_overhead()*WHISPER_MAX_NODES + ggml_graph_overhead();
+    static std::vector<uint8_t> buf(buf_size);
+
     struct ggml_init_params params = {
-        /*.mem_size   =*/ wstate.alloc_encode.meta.size(),
-        /*.mem_buffer =*/ wstate.alloc_encode.meta.data(),
+        /*.mem_size   =*/ buf_size,
+        /*.mem_buffer =*/ buf.data(),
         /*.no_alloc   =*/ true,
     };
 
     struct ggml_context * ctx0 = ggml_init(params);
 
     ggml_cgraph * gf = ggml_new_graph_custom(ctx0, WHISPER_MAX_NODES, false);
 
-    ggml_allocr * alloc = wstate.alloc_encode.alloc;
-
     struct ggml_tensor * KQscale = ggml_new_tensor_1d(ctx0, GGML_TYPE_F32, 1);
-    ggml_allocr_alloc(alloc, KQscale);
+    //ggml_allocr_alloc(alloc, KQscale);
 
-    if (!ggml_allocr_is_measure(alloc)) {
-        ggml_set_f32(KQscale, 1.0f/sqrt(float(n_state)/n_head));
-    }
+    //if (!ggml_allocr_is_measure(alloc)) {
+    //    ggml_set_f32(KQscale, 1.0f/sqrt(float(n_state)/n_head));
+    //}
 
     struct ggml_tensor * cur = ggml_view_tensor(ctx0, wstate.embd_conv);
```

Comment:
@slaren

I'm migrating `whisper.cpp` to use the new scheduler and backend API.

Here, if I want to use the result of a previous graph (`wstate.embd_conv`) as an input, what would be the proper way to do it? If I allocated `wstate.embd_conv` in a dedicated input buffer, then do I need to change the following:

https://github.com/ggerganov/whisper.cpp/blob/005b8ccbf0288975985ef15082579b34cadbf36a/whisper.cpp#L1592-L1593

to `ggml_cpy(ctx0, cur, wstate.embd_conv);`? If yes, is there a way to avoid the copy?

Reply:
It is fine to use a pre-allocated tensor, I don't think that you need to make a copy here. Conceptually this should be the same as using a weight pre-allocated in another buffer, right? Maybe I am missing something.

Reply:
To elaborate more: when using `ggml-backend` directly without `ggml-backend-sched`, all the tensors must be allocated in the same backend. If you wanted to evaluate a graph in a backend, and then use the output of the graph as in input of a different graph evaluated on a different backend, you would need to use `ggml_backend_tensor_copy` to copy the tensor to the other backend. When using `ggml-backend-sched`, these copies should be done automatically.

Reply:
>  Conceptually this should be the same as using a weight pre-allocated in another buffer, right?

Yes. Currently it errors that `wstate.embd_conv` does not have a backend, but I think I see the problem. Will try to fix it

Another question: here I am creating a separate scheduler for each separate graph:

https://github.com/ggerganov/whisper.cpp/blob/005b8ccbf0288975985ef15082579b34cadbf36a/whisper.cpp#L2865-L2899

Is there a way to do this with a single scheduler?

Reply:
No, as it is currently, each graph must use a different scheduler. Maybe this is something that we could look into adding in the future, but currently all the allocations happen during the call to `ggml_backend_sched_init_measure`, and there is no way to add other graphs.